sql: Add support for trivial column type changes.

bobvawter opened the pull request:
This change enables column type changes for situations in which the on-disk
encoding of the value does not change between the source and target types.

The current code is a bit over-generalized for the current implementation, but
is structured the way it is to more easily support the next round of
development for generalized type conversions.

Release note: None

bobvawter committed:
sql: Add support for trivial column type changes.

This change enables column type changes for situations in which the on-disk
encoding of the value does not change between the source and target types.

The current code is a bit over-generalized for the current implementation, but
is structured the way it is to more easily support the next round of
development for generalized type conversions.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25777)
<!-- Reviewable:end -->

jordanlewis commented:
Excuse my drive-by comment, but I have a question for you: does your implementation allow converting between our column type aliases? I was just thinking about these a little and your PR title reminded me of the issue, so I figured I'd ask.

Specifically, we've got a couple of column types that for all intents and purposes are identical - except their names are preserved in the `VisibleType` field of the `ColumnDescriptor`. For example, `CREATE TABLE a (a INT PRIMARY KEY)` and `CREATE TABLE b (a INTEGER PRIMARY KEY)` are identical except for the fact that `SHOW CREATE TABLE` on each returns the precise input type name.

I'd just like to make sure that any alter tables correctly change the name of the column type, despite the fact that there's no "actual" change going on under the hood. 

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-:-LD3N8-hA1J7IV8xLpe-:b-kiivg3)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
@jordanlewis those kinds of visible-type-only changes haven't come up in discussion so far, but they should only be a couple of lines to make sure that it works.  I'll add a test case for that tomorrow.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-:-LD3U3-jBC42e4VOQX3s:b-53q5t2)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I would also like to see a buncha logic tests (can be a big table with lots of types) with a SHOW CREATE TABLE at the end to verify everything got done correctly from a table desc perspective.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/alter_column_type.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3Xqq8E4oQbDn6Ve3Y:-LD3Xqq8E4oQbDn6Ve3Z:bueh4tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L28)):*
> ```Go
> 
> // These columnConversionKinds are ordered from easiest to hardest.
> // TODO: Should this be a type hierarchy to support conversion planning?
> ```

any TODOs should be of the form `TODO(yourname): Notes ...`

---

*[pkg/sql/alter_column_type.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4X8VOCLl_rrG8F4bV:-LD4X8VOCLl_rrG8F4bW:b-619zuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L69)):*
> ```Go
> }
> 
> func (k columnConversionKind) String() string {
> ```

add this line: `//go:generate stringer -type=columnConversionKind -trimprefix columnConversion`, run `go generate` in this directory, and you'll get a magic string func

---

*[pkg/sql/alter_column_type.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4VkFB1UnKe4bo0D1I:-LD4VkFB1UnKe4bo0D1J:b-3yf7as) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L94)):*
> ```Go
> 	sqlbase.ColumnType_FLOAT: {
> 		// Floats are always encoded as 64-bit values on disk
> 		sqlbase.ColumnType_FLOAT: classifierHardestOf(classifierPrecision, classifierWidth),
> ```

neither postgres nor cockroach support scale (width) on floats. while postgres supports precision, it only means double or single precision. and cockroach doesn't care about the float precision at all, so i think all float -> float conversions are trivial

---

*[pkg/sql/alter_column_type.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4XceU3zo5sSDNEOBq:-LD4XceU3zo5sSDNEOBr:bbv8nxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L97)):*
> ```Go
> 	},
> 	sqlbase.ColumnType_INT: {
> 		sqlbase.ColumnType_INT: classifierWidth,
> ```

i'm not sure this will work correctly for `bit(n)` conversions. see https://www.cockroachlabs.com/docs/stable/int.html

---

*[pkg/sql/alter_column_type.go, line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3YRinBvMsWn6dxz6P:-LD3YRinBvMsWn6dxz6Q:b-932gwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L103)):*
> ```Go
> 		// bytes type has an unlimited width or that we have at least
> 		// 4x the number of bytes as known-maximum characters.
> 		sqlbase.ColumnType_BYTES: func(string *sqlbase.ColumnType, bytes *sqlbase.ColumnType) columnConversionKind {
> ```

we generally don't use keywords as variable names. could you change string to something else?

---

*[pkg/sql/alter_column_type.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4RKtXBXYcq-ea_yf1:-LD4RKtXBXYcq-ea_yf2:bngyzly) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L211)):*
> ```Go
> 
> 	// Use a placeholder just to sub in the original type.
> 	fromPlaceholder, err := (&tree.Placeholder{}).TypeCheck(&ctx, oldType.ToDatumType())
> ```

Make sure knz or someone from the sql team OKs this. It's clever but I want to make sure it's the best we can do given our current API.

---

*[pkg/sql/alter_column_type_test.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4XqVTC32EXLr8yMcu:-LD4XqVTC32EXLr8yMcv:bpnm0ew) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type_test.go#L36)):*
> ```Go
> 
> // testKey exists because sqlbase.ColumnType isn't map-compatible
> type testKey struct {
> ```

i'd prefer all these types and columnCovresionInfo to be moved to a single function (i think you can merge the two test functions, also). you'll have to make a helper for toColumnType, but this will keep the namespace a bit cleaner.

---

*[pkg/sql/alter_column_type_test.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4UV7j4uZpnYK10k_K:-LD4UV7j4uZpnYK10k_L:b35jgcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type_test.go#L57)):*
> ```Go
> var columnConversionInfo testMap = map[testKey]map[testKey]columnConversionKind{
> 	{SemanticType: sqlbase.ColumnType_BYTES}: {
> 		{SemanticType: sqlbase.ColumnType_STRING}:            columnConversionValidate,
> ```

need to add some noop tests

---

*[pkg/sql/alter_table.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3U1ARCeb3No-6qi-2:-LD3U1ARCeb3No-6qi-3:bckdxwy) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_table.go#L23)):*
> ```Go
> 	"fmt"
> 
> 	"github.com/cockroachdb/cockroach/pkg/sql/coltypes"
> ```

you may need to update your copy of goimports because it should have group this correctly

---

*[pkg/sql/alter_table.go, line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3T2YW87SQD0kxxC_O:-LD3T2YW87SQD0kxxC_P:b8qivri) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_table.go#L677)):*
> ```Go
> 
> 		// Special handling for STRING COLLATE xy to verify that we recognize the language.
> 		if len(t.Collation) > 0 {
> ```

t.Collation != ""

---

*[pkg/sql/alter_table.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3Tduo4AaLgGkQX3XO:-LD3Tduo4AaLgGkQX3XP:bv86jxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_table.go#L680)):*
> ```Go
> 			if types.IsStringType(datum) {
> 				if _, err := language.Parse(t.Collation); err != nil {
> 					return pgerror.NewErrorf(pgerror.CodeSyntaxError, `invalid COLLATE locale: "%s""`, t.Collation)
> ```

`invalid locale %s` is the error message we use elsewhere for this. i think we can keep it as that.

---

*[pkg/sql/parser/sql.y, line 1503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3Seot5gePLy47Ot7Y:-LD3Seou1dOKmxvOGT1k:b-69zxpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/parser/sql.y#L1503)):*
> ```Yacc
> opt_alter_column_collate:
>   COLLATE collation_name { $$ = $2 }
> | /* EMPTY */ { $$ = "" }
> ```

why is this necessary? is depending on the zero value for this not enough? did its omission break something we should have had a test for?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-:-LD4YTKN-FquG_hb5t8Y:b4z86ri)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
PTAL:   Added a visible-type-only test and now verifying the `ALTER COLUMN TYPE` for each of the conversion sub-tests.

---

Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/sql/alter_column_type.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3Xqq8E4oQbDn6Ve3Y:-LD6tb1z8q-HpKRMvU1v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L28)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

any TODOs should be of the form `TODO(yourname): Notes ...`
</blockquote></details>

Done.

---

*[pkg/sql/alter_column_type.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4X8VOCLl_rrG8F4bV:-LD6tdCj0FBw0CeSpUkO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L69)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

add this line: `//go:generate stringer -type=columnConversionKind -trimprefix columnConversion`, run `go generate` in this directory, and you'll get a magic string func
</blockquote></details>

Done.

---

*[pkg/sql/alter_column_type.go, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4VkFB1UnKe4bo0D1I:-LD6vQrDAnlhkHoeW_Np:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L94)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

neither postgres nor cockroach support scale (width) on floats. while postgres supports precision, it only means double or single precision. and cockroach doesn't care about the float precision at all, so i think all float -> float conversions are trivial
</blockquote></details>

Done.

---

*[pkg/sql/alter_column_type.go, line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4XceU3zo5sSDNEOBq:-LD7jGQy5B1CYEjbsfy9:b-dae0lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L97)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

i'm not sure this will work correctly for `bit(n)` conversions. see https://www.cockroachlabs.com/docs/stable/int.html
</blockquote></details>

The bit types are kind of wonky to begin with.  I'm not sure we want to support converting between `INT <--> BIT` in the current state without a `USING expression`.  See https://github.com/cockroachdb/cockroach/issues/20991 for discussion.  For the moment, I'm going to have the code punt on this and we can sync up with the sql team on the state of `BIT` compatibility.

---

*[pkg/sql/alter_column_type.go, line 103 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3YRinBvMsWn6dxz6P:-LD6vrfJ751-Lt2IT1Ic:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L103)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

we generally don't use keywords as variable names. could you change string to something else?
</blockquote></details>

Done.

---

*[pkg/sql/alter_column_type.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4RKtXBXYcq-ea_yf1:-LD6vueaCkhQUHShGVYJ:bkw3i4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type.go#L211)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Make sure knz or someone from the sql team OKs this. It's clever but I want to make sure it's the best we can do given our current API.
</blockquote></details>

I had a conversation with knz about this in #sql the other day.  I'll ping him to take a look at this.

---

*[pkg/sql/alter_column_type_test.go, line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4XqVTC32EXLr8yMcu:-LD6xbbp3a3rH0qoSYHC:b-v3xv0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type_test.go#L36)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

i'd prefer all these types and columnCovresionInfo to be moved to a single function (i think you can merge the two test functions, also). you'll have to make a helper for toColumnType, but this will keep the namespace a bit cleaner.
</blockquote></details>

Restructured with nested `t.Run()` calls to wrap this up in a private namespace.

---

*[pkg/sql/alter_column_type_test.go, line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD4UV7j4uZpnYK10k_K:-LD7jBwvE5wuu4Vkqb-5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_column_type_test.go#L57)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

need to add some noop tests
</blockquote></details>

Done.

---

*[pkg/sql/alter_table.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3U1ARCeb3No-6qi-2:-LD6yRcW1nQTZJPRIFlg:b-ba3sod) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_table.go#L23)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

you may need to update your copy of goimports because it should have group this correctly
</blockquote></details>

Even with the latest goimports, I still needed to move this line into the appropriate imports group in order to have it sorted correctly.

---

*[pkg/sql/alter_table.go, line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3T2YW87SQD0kxxC_O:-LD6zZWO4-_klE4dBy3l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_table.go#L677)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

t.Collation != ""
</blockquote></details>

Done.

---

*[pkg/sql/alter_table.go, line 680 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3Tduo4AaLgGkQX3XO:-LD6zakv27GWNJIeX-ZG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/alter_table.go#L680)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

`invalid locale %s` is the error message we use elsewhere for this. i think we can keep it as that.
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 1503 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LD3Seot5gePLy47Ot7Y:-LD6s7670KqhCznD11j5:bb6isnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/57719950145066918885d3bd3975d16f0207bdc8/pkg/sql/parser/sql.y#L1503)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

why is this necessary? is depending on the zero value for this not enough? did its omission break something we should have had a test for?
</blockquote></details>

Without this, I was getting the next token in the input as the value of `opt_alter_column_collate`.  For instance, when parsing `colname BIT(5)`, I'd get a left parenthesis in the locale field.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-:-LD7jwIc0tDxyzn5a-i_:big4yrs)*
<!-- Sent from Reviewable.io -->

mjibson commented:
I'm going to wait for the logic tests before I do a deeper look because I think we can move a fair amount of code there.

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/alter_column_type_test.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDA2LSsEvxunyhfF3Ev:-LDA2LSsEvxunyhfF3Ew:b-7l1rz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type_test.go#L39)):*
> ```Go
> // column's visible type e.g. INT -> INTEGER, which is otherwise a
> // no-op behavior.
> func TestVisibleTypeChangeOnly(t *testing.T) {
> ```

this can be a logic test

---

*[pkg/sql/alter_table.go, line 680 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDA2YemD6CmV885O-xR:-LDA2YemD6CmV885O-xS:b-q3knpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_table.go#L680)):*
> ```Go
> 			if types.IsStringType(datum) {
> 				if _, err := language.Parse(t.Collation); err != nil {
> 					return pgerror.NewErrorf(pgerror.CodeSyntaxError, `invalid locale: "%s""`, t.Collation)
> ```

still not identical. remove the colon and quotes.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-:-LDA2qZL0DVE2mLxrL-m:b-msbk3a)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/alter_column_type.go, line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDCf7RdDm2w6uXhtWdD:-LDCf7RdDm2w6uXhtWdE:bblllm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> 
> package sql
> ```

I think we should create a new package schemachange under pkg/sql/schemachange and put this code in there.  

---

*[pkg/sql/alter_column_type.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDCbxfFE_vrvNnKBpru:-LDCbxfFE_vrvNnKBprv:b-pfrvn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type.go#L41)):*
> ```Go
> 	// columnConversionNoOp indicates that the requested change is
> 	// meaningless.
> 	columnConversionNoOp
> ```

Didn't quite understand where one might use this one

---

*[pkg/sql/alter_column_type.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDCdYd39QGSKU27Sopu:-LDCdYd44icYfND9q4FE:baraml7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type.go#L72)):*
> ```Go
> // classifiers contains the logic for looking up conversions which
> // don't require a fully-generalized approach.
> var classifiers = map[sqlbase.ColumnType_SemanticType]map[sqlbase.ColumnType_SemanticType]classifier{
> ```

wonder if you can use const here. hoping :-)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
PTAL.  I want to keep the main `TestColumnConversions()` function as currently structured until I see how the conversion matrix shakes out.  I'm worried about it being easy to miss conversion cases using logic tests and also want to cross-check the `columnConversionInfo` data against the actual system.

---

Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/alter_table.go, line 680 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDA2YemD6CmV885O-xR:-LDDSLuzCIdnI3CVFW5w:b20nmkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_table.go#L680)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

still not identical. remove the colon and quotes.
</blockquote></details>

Whoops.  Fixed.

---

*[pkg/sql/alter_column_type.go, line 15 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDCf7RdDm2w6uXhtWdD:-LDDSNnL3qFZGq4iu5e8:b-ohgtl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type.go#L15)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think we should create a new package schemachange under pkg/sql/schemachange and put this code in there.  
</blockquote></details>

Moved to new package.

---

*[pkg/sql/alter_column_type.go, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDCbxfFE_vrvNnKBpru:-LDDSPzF4qk6MAhVTI4q:b9id3rl) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type.go#L41)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Didn't quite understand where one might use this one
</blockquote></details>

I have removed the NoOp value in favor of returning Trivial.  The doc on `ClassifyConversion()` has been updated to indicate that it's the caller's responsibility to identify equal (or equivalent) types.

---

*[pkg/sql/alter_column_type.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDCdYd39QGSKU27Sopu:-LDDSiIZ4wx6fxtqm2ie:b-cx1nlu) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type.go#L72)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

wonder if you can use const here. hoping :-)
</blockquote></details>

Maps aren't `const`-compatible, are they?

---

*[pkg/sql/alter_column_type_test.go, line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDA2LSsEvxunyhfF3Ev:-LDDSshVBPvRAUfh2cDg:b-knrz35) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c06647379d1c7f8686f348785f57db54fc00a28/pkg/sql/alter_column_type_test.go#L39)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

this can be a logic test
</blockquote></details>

Moved `TestVisibleTypeChangeOnly` to a logic test, added some other error-condition checks there, too.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-:-LDDTPXcB6oPDwkack6T:b-8r7f5i)*
<!-- Sent from Reviewable.io -->

vivekmenezes reviewed (commented):
LGTM

mjibson commented on pkg/sql/logictest/testdata/logic_test/alter_column_type:
```diff
@@ -0,0 +1,65 @@
+# LogicTest: default parallel-stmts distsql distsql-metadata
```
I want a test of normal stuff that is hard coded to something we know. Can you add this test:

```
statement ok
CREATE TABLE t (s STRING, sl STRING(5), t TIME, ts TIMESTAMP)

statement ok
INSERT INTO t VALUES ('some string', 'short', TIME '05:04:00', now())

statement ok
ALTER TABLE t ALTER s TYPE BYTES, ALTER sl TYPE STRING(6), ALTER t TYPE TIMETZ, ALTER ts TYPE TIMESTAMPTZ)

query TT
SHOW CREATE TABLE t
----
<whatever>

query TTTT
SELECT * FROM t
----
<whatever>
```

bobvawter commented:
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/alter_column_type, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25777#-LDDS8LQ6JSvrK4jGx76-r3-1:-LDDeQ9_5B_1BK48yrE-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/08daf91ee9030e49d096b4c55679146348f9bcad/pkg/sql/logictest/testdata/logic_test/alter_column_type#L1)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I want a test of normal stuff that is hard coded to something we know. Can you add this test:

```
statement ok
CREATE TABLE t (s STRING, sl STRING(5), t TIME, ts TIMESTAMP)

statement ok
INSERT INTO t VALUES ('some string', 'short', TIME '05:04:00', now())

statement ok
ALTER TABLE t ALTER s TYPE BYTES, ALTER sl TYPE STRING(6), ALTER t TYPE TIMETZ, ALTER ts TYPE TIMESTAMPTZ)

query TT
SHOW CREATE TABLE t
----
<whatever>

query TTTT
SELECT * FROM t
----
<whatever>
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25777)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
bors r+

bobvawter commented:
Build flake, retrying.

bors r+

bobvawter commented:
Re-synced with master to fix build break.

bors r+

jordanlewis commented:
@bobvawter @vivekmenezes this needed a release note - can you please work with @lhirata to get this documented?
